### PR TITLE
Improve runner performance

### PR DESCRIPTION
**Pull request by jamestalmage**
The main premise I had while creating this: Deferring when you don't need to is expensive.

 This rewrites the runner and test to use the following semantics:

If a test never uses an async features (i.e. `cb` mode, `t.throws(promise)`, return a `promise` / `observable`) the next test is executed synchronously.

This requires a unified API between tests and the runner:

```js
// returns a "result" or promise for a result
var result = runnable.run();

// a "result" has the following properties:
result = {
  passed: true || false,
  reason: Error,
  result: "some result"
}
```

Because `run` can return either a result or a promise for a result, you frequently need to check which it is using `is-promise`. This leads to some verbose code, but it runs faster, and is fairly well encapsulated in the `sequence` and `concurrent` classes.

AVA master:
```
ava --verbose  9.38s user 0.63s system 461% cpu 2.173 total
```

This PR:
```
ava --verbose  8.46s user 0.65s system 522% cpu 1.744 total
```

### CODE REVIEW

**Comment by jamestalmage** [comment]
All working now, using benchmarks from [jamestalmage/emoji-aware#no-assertion-plans](https://github.com/jamestalmage/emoji-aware/tree/no-assertion-plans):

Master:
```
ava --verbose  6.86s user 0.56s system 439% cpu 1.689 total
ava --verbose  6.78s user 0.55s system 438% cpu 1.669 total
ava --verbose  6.73s user 0.55s system 432% cpu 1.681 total
```

This PR:
```
ava --verbose  5.25s user 0.53s system 432% cpu 1.337 total
ava --verbose  4.97s user 0.50s system 419% cpu 1.303 total
ava --verbose  5.05s user 0.50s system 426% cpu 1.301 total
```

**Comment by jamestalmage** [comment]
Using the following benchmark (no assertions - just tracking runner performance):

```js
import test from './';

for (var i = 0; i < 10000; i++) {
	test('test' + i, t => {});
}
```

Master:
```
node ./cli.js benchmark.js --verbose  3.15s user 0.31s system 104% cpu 3.318 total
node ./cli.js benchmark.js --verbose  2.76s user 0.28s system 103% cpu 2.937 total
node ./cli.js benchmark.js --verbose  2.73s user 0.28s system 103% cpu 2.911 total
```

This PR:
```
node ./cli.js benchmark.js --verbose  1.44s user 0.25s system 107% cpu 1.573 total
node ./cli.js benchmark.js --verbose  1.40s user 0.25s system 106% cpu 1.544 total
node ./cli.js benchmark.js --verbose  1.38s user 0.24s system 108% cpu 1.499 total
```

**Comment by jamestalmage** [comment]
This also seems to provide a slight performance boost for async tests:

```js
import test from './';

for (var i = 0; i < 10000; i++) {
	test('test' + i, t => {return Promise.resolve()});
}
```

Master:

```
node ./cli.js benchmark.js --verbose  2.84s user 0.28s system 103% cpu 3.010 total
node ./cli.js benchmark.js --verbose  2.86s user 0.28s system 103% cpu 3.033 total
node ./cli.js benchmark.js --verbose  2.82s user 0.28s system 104% cpu 2.980 total
```

This PR:
```
node ./cli.js benchmark.js --verbose  2.68s user 0.28s system 103% cpu 2.843 total
node ./cli.js benchmark.js --verbose  2.69s user 0.28s system 104% cpu 2.854 total
node ./cli.js benchmark.js --verbose  2.72s user 0.27s system 104% cpu 2.880 total
```

**Comment by jamestalmage** [comment]
Mixed sync/async:
```
import test from './';

test.beforeEach(t => {});
test.afterEach(t => {});

for (var i = 0; i < 10000; i++) {
	test('test' + i, t => {return Promise.resolve()});
}
```

Master:
```
node ./cli.js benchmark.js --verbose  4.78s user 0.43s system 102% cpu 5.075 total
node ./cli.js benchmark.js --verbose  4.64s user 0.42s system 102% cpu 4.953 total
node ./cli.js benchmark.js --verbose  4.75s user 0.43s system 102% cpu 5.041 total
```

This PR:
```
node ./cli.js benchmark.js --verbose  3.56s user 0.38s system 103% cpu 3.797 total
node ./cli.js benchmark.js --verbose  3.56s user 0.38s system 103% cpu 3.791 total
node ./cli.js benchmark.js --verbose  3.62s user 0.39s system 103% cpu 3.888 total
```

File: lib/concurrent.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+'use strict';
+var Promise = require('bluebird');
+var isPromise = require('is-promise');
+var BAIL_ERROR = new Error();
+
+module.exports = Concurrent;
+
+function Concurrent(tests, bail) {
+	if (!this instanceof Concurrent) {
+		throw new Error('must use `new Concurrent`');
```
Comment (sindresorhus) [inline lib/concurrent.js]: Should be `TypeError` to match ES2015 classes and message: `Class constructor Concurrent cannot be invoked without 'new'`

File: lib/concurrent.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	var ret = Promise.all(results).then(function (results) {
+		return {
+			passed: passed,
+			reason: reason,
+			result: results
+		};
+	});
+
+	if (bail) {
+		ret = ret.catch(function (e) {
```
Comment (sindresorhus) [inline lib/concurrent.js]: `e` => `err`

File: lib/sequence.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
+'use strict';
+var Promise = require('bluebird');
+var isPromise = require('is-promise');
+
+module.exports = Sequence;
+
+function Sequence(tests, bail) {
+	if (!this instanceof Sequence) {
+		throw new Error('Sequence must be called with new');
```
Comment (sindresorhus) [inline lib/sequence.js]: Same as https://github.com/sindresorhus/ava/pull/466/files#r50755945

File: lib/test-collection.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		if (metadata.exclusive && !this.hasExclusive) {
+			this.hasExclusive = true;
+			this.serial = [];
+			this.concurrent = [];
+		}
+		(metadata.serial ? this.serial : this.concurrent).push(test);
+		return;
+	}
+	if (metadata.exclusive) {
+		throw new Error('you can\'t use only with a ' + type + ' test');
```
Comment (sindresorhus) [inline lib/test-collection.js]: I would quote `only` to make it clearer that it's not a normal word.

File: test/hooks.js [hunk 5]
```diff
@@ -257,12 +257,12 @@ test('shared context', function (t) {
 	var runner = new Runner();
 
 	runner.before(function (a) {
-		a.is(a.context, undefined);
-		a.context = {arr: []};
+		a.is(a.context, null);
+		// a.context = {arr: []};
```
Comment (sindresorhus) [inline test/hooks.js]: ?

File: test/runner.js [hunk 6]
```diff
@@ -130,7 +130,7 @@ test('methods are chainable: serial.only', function (t) {
 	t.true(runner.tests[0].metadata.exclusive);
 	t.false(runner.tests[0].metadata.skipped);
 	t.end();
-});
+});   */
```
Comment (sindresorhus) [inline test/runner.js]: :ghost:

File: lib/test.js [hunk 7]
```diff
@@ -11,12 +11,13 @@ var assert = require('./assert');
 var enhanceAssert = require('./enhance-assert');
 var globals = require('./globals');
 
-function Test(title, fn) {
+function Test(title, fn, contextRef, report) {
 	if (!(this instanceof Test)) {
-		return new Test(title, fn);
+		return new Test(title, fn, contextRef);
```
Comment (matthewbauer) [inline lib/test.js]: this should be return new Test(title, fn, contextRef, report);

File: test/hooks.js (hunk 5 above)
Comment (jamestalmage) [inline test/hooks.js]: Setting `t.context` in `before` or `after` will now throw an error. Previously we silently let them set it, but it was not passed to future tests.

File: test/runner.js (hunk 6 above)
Comment (jamestalmage) [inline test/runner.js]: oops. Yeah that should just be deleted. It's well covered in `test-collection`

File: lib/test.js (hunk 7 above)
Comment (jamestalmage) [inline lib/test.js]: True. (mistakes like this is why I prefer to throw instead of doing magic).

File: test/hooks.js (hunk 5 above)
Comment (jamestalmage) [inline test/hooks.js]: I probably should add a test covering that.

**Comment by matthewbauer** [comment]
Is there a way to get nested groups (#222) working in this? I'm thinking of an implementation like this:

```js
test.group(function(test) {
  test("test a", function() {
  });
  test.beforeEach("before test a but not test b", function() {
  });
});
test("test b", function() {});
```

Basically test.group just generates another runner that is added into the Runner.prototype.tests array. All of the hooks work as expected, just with the new test in context.

I was working on a PR to do this but then I saw the changes here and wanted to check what your thoughts are before moving forward.

**Comment by sindresorhus** [comment]
:clap: From a quick skim this looks very promising, although a lot to take in. I like the improved separation of concerns and performance improvements, but not totally happy about the added verboseness. Would be nice if we could follow up later with ways to simplify it. I'll try to do a more thorough review tomorrow.

-

@vdemedes @ariporad @sotojuan @novemberborn I could use some help reviewing this :)

**Comment by sindresorhus** [comment]
@matthewbauer Can you move your comment to https://github.com/sindresorhus/ava/issues/222 so not to derail this PR discussion? :)

**Comment by matthewbauer** [comment]
:ok_hand:

**Comment by jamestalmage** [comment]
> Would be nice if we could follow up later with ways to simplify it.

Agreed. Getting a good benchmarking suite (#460) would be really helpful as well.

**Comment by vdemedes** [comment]
Massive applause for huge amount of work, @jamestalmage.

I have some style/cleanup-related suggestions, but I can follow up with a PR myself ;)

**Comment by sindresorhus** [comment]
> I have some style/cleanup-related suggestions, but I can follow up with a PR myself ;)

Yes, let's do style stuff later. (I did a few myself, but realized it's better not to dilute the review with nitpicks).

**Comment by jamestalmage** [comment]
Once merged, we should all dogfood this on existing projects before release, (maybe deploy with a `next` tag for a while as well). There are enough changes here that our regression exposure is pretty large. I've added pretty exhaustive tests, but we should still thoroughly vet it.

**Comment by sindresorhus** [comment]
> Once merged, we should all dogfood this on existing projects before release

:+1: I've already tried it on `got` and `pageres` and seems to work good.

> maybe deploy with a next tag for a while as well

:+1:

**Comment by sotojuan** [comment]
Currently in class but from what I skimmed it looks good :-) Awesome work!

**Comment by sindresorhus** [comment]
Landed! Awesome work @jamestalmage :)

unicorn17

**Pull request closed without merging**